File: tests/FakeServer2.java
import com.sun.net.httpserver.HttpExchange;
import com.sun.net.httpserver.HttpServer;

import java.io.ByteArrayOutputStream;
import java.io.OutputStream;
import java.net.InetSocketAddress;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.Executors;

public class FakeServer2 {
    private static final Path ROOT = Path.of("tests", "fake_server2_inbox");
    private static final Map<String, Map<Integer, byte[]>> CHUNKS = new ConcurrentHashMap<>();
    private static final Map<String, String> META = new ConcurrentHashMap<>();

    public static void main(String[] args) throws Exception {
        Files.createDirectories(ROOT);
        HttpServer server = HttpServer.create(new InetSocketAddress("127.0.0.1", 19081), 0);
        server.createContext("/health", ex -> json(ex, 200,
                "{\"status\":\"ok\",\"role\":\"processing_storage_readonly_output\"}"));
        server.createContext("/internal/uploads", FakeServer2::uploads);
        server.createContext("/internal/tasks", FakeServer2::tasks);
        server.createContext("/internal/results", FakeServer2::results);
        server.setExecutor(Executors.newCachedThreadPool());
        server.start();
        System.out.println("fake server2 listening on http://127.0.0.1:19081");
    }

    private static void uploads(HttpExchange ex) throws java.io.IOException {
        String path = ex.getRequestURI().getPath();
        String method = ex.getRequestMethod();
        byte[] body = ex.getRequestBody().readAllBytes();
        if ("POST".equals(method) && "/internal/uploads".equals(path)) {
            META.put("init", new String(body, StandardCharsets.UTF_8));
            CHUNKS.put("up-live", new ConcurrentHashMap<>());
            json(ex, 200, "{\"upload_id\":\"up-live\",\"chunk_size\":8388608}");
            return;
        }
        if ("PUT".equals(method) && path.contains("/chunks/")) {
            int idx = Integer.parseInt(path.substring(path.lastIndexOf('/') + 1));
            CHUNKS.computeIfAbsent("up-live", k -> new ConcurrentHashMap<>()).put(idx, body);
            json(ex, 200, "{\"ok\":true}");
            return;
        }
        if ("POST".equals(method) && path.endsWith("/complete")) {
            Map<Integer, byte[]> parts = CHUNKS.getOrDefault("up-live", Map.of());
            ByteArrayOutputStream assembled = new ByteArrayOutputStream();
            for (int i = 0; i < parts.size(); i++) {
                assembled.write(parts.getOrDefault(i, new byte[0]));
            }
            String init = META.getOrDefault("init", "{}");
            String fileId = extract(init, "file_id");
            String name = extract(init, "file_name");
            if (fileId.isBlank()) fileId = "unknown";
            if (name.isBlank()) name = "payload.bin";
            Path dir = ROOT.resolve(fileId);
            Files.createDirectories(dir);
            Path dest = dir.resolve(name);
            Files.write(dest, assembled.toByteArray());
            Files.writeString(dir.resolve("meta.json"), init, StandardCharsets.UTF_8);
            String source = dest.toAbsolutePath().toString().replace('\\', '/');
            json(ex, 200, "{\"source_path\":\"" + source + "\",\"bytes\":" + assembled.size() + "}");
            return;
        }
        json(ex, 404, "{\"detail\":\"not found\"}");
    }

    private static void tasks(HttpExchange ex) throws java.io.IOException {
        byte[] body = ex.getRequestBody().readAllBytes();
        if ("POST".equals(ex.getRequestMethod()) && "/internal/tasks".equals(ex.getRequestURI().getPath())) {
            Files.write(ROOT.resolve("last_task.json"), body);
            json(ex, 202, "{\"task_id\":\"task-live\",\"status\":\"COMPLETED\",\"result_id\":\"res-live\"}");
            return;
        }
        if ("GET".equals(ex.getRequestMethod())) {
            json(ex, 200, "{\"task_id\":\"task-live\",\"status\":\"COMPLETED\",\"result_id\":\"res-live\"}");
            return;
        }
        json(ex, 404, "{\"detail\":\"not found\"}");
    }

    private static void results(HttpExchange ex) throws java.io.IOException {
        String path = ex.getRequestURI().getPath();
        if (path.endsWith("/approve")) {
            json(ex, 200, "{\"result_id\":\"res-live\",\"approval_status\":\"APPROVED\"}");
            return;
        }
        if (path.endsWith("/trace")) {
            json(ex, 200, "{\"result_id\":\"res-live\",\"trace\":\"ok\"}");
            return;
        }
        json(ex, 404, "{\"detail\":\"not found\"}");
    }

    private static void json(HttpExchange ex, int status, String body) throws java.io.IOException {
        byte[] bytes = body.getBytes(StandardCharsets.UTF_8);
        ex.getResponseHeaders().set("Content-Type", "application/json");
        ex.sendResponseHeaders(status, bytes.length);
        try (OutputStream os = ex.getResponseBody()) {
            os.write(bytes);
        }
    }

    private static String extract(String json, String field) {
        String key = "\"" + field + "\"";
        int idx = json.indexOf(key);
        if (idx < 0) return "";
        int colon = json.indexOf(':', idx);
        int start = json.indexOf('"', colon + 1) + 1;
        int end = json.indexOf('"', start);
        return json.substring(start, end);
    }
}

File: tests/fake_server2.py
"""Minimal Server2 stand-in used by Server1 Java dispatch self-test.

Implements the same calls the Java client makes:
health, chunked upload, task submit/status, result approve/trace.
"""
from __future__ import annotations

import hashlib
import json
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from pathlib import Path

ROOT = Path(__file__).resolve().parent / "fake_server2_inbox"
ROOT.mkdir(parents=True, exist_ok=True)
CHUNKS: dict[str, dict[int, bytes]] = {}
META: dict[str, dict] = {}


class Handler(BaseHTTPRequestHandler):
    def log_message(self, fmt: str, *args) -> None:
        print("[fake-server2]", fmt % args)

    def _send(self, code: int, payload: dict) -> None:
        raw = json.dumps(payload).encode("utf-8")
        self.send_response(code)
        self.send_header("Content-Type", "application/json")
        self.send_header("Content-Length", str(len(raw)))
        self.end_headers()
        self.wfile.write(raw)

    def do_GET(self) -> None:
        if self.path == "/health":
            self._send(200, {"status": "ok", "role": "processing_storage_readonly_output"})
            return
        if self.path.startswith("/internal/tasks/"):
            self._send(200, {"task_id": "task-live", "status": "COMPLETED", "result_id": "res-live"})
            return
        if self.path.endswith("/trace"):
            self._send(200, {"result_id": "res-live", "trace": "ok"})
            return
        self._send(404, {"detail": "not found"})

    def do_POST(self) -> None:
        length = int(self.headers.get("Content-Length", "0"))
        body = self.rfile.read(length)
        if self.path == "/internal/uploads":
            data = json.loads(body.decode("utf-8") or "{}")
            upload_id = "up-live"
            META[upload_id] = data
            CHUNKS[upload_id] = {}
            self._send(200, {"upload_id": upload_id, "chunk_size": 8 * 1024 * 1024})
            return
        if self.path.endswith("/complete"):
            upload_id = self.path.split("/")[3]
            blob = b"".join(CHUNKS.get(upload_id, {}).get(i, b"") for i in range(len(CHUNKS.get(upload_id, {}))))
            meta = META.get(upload_id, {})
            file_id = meta.get("file_id", "unknown")
            name = meta.get("file_name", "payload.bin")
            dest_dir = ROOT / file_id
            dest_dir.mkdir(parents=True, exist_ok=True)
            dest = dest_dir / name
            dest.write_bytes(blob)
            (dest_dir / "meta.json").write_text(json.dumps({
                **meta,
                "sha256": hashlib.sha256(blob).hexdigest(),
                "bytes": len(blob),
                "source_path": str(dest),
            }, ensure_ascii=False, indent=2), encoding="utf-8")
            self._send(200, {"source_path": str(dest), "sha256": hashlib.sha256(blob).hexdigest()})
            return
        if self.path == "/internal/tasks":
            data = json.loads(body.decode("utf-8") or "{}")
            (ROOT / "last_task.json").write_text(json.dumps(data, ensure_ascii=False, indent=2), encoding="utf-8")
            self._send(202, {"task_id": "task-live", "status": "COMPLETED", "result_id": "res-live"})
            return
        if self.path.endswith("/approve"):
            self._send(200, {"result_id": "res-live", "approval_status": "APPROVED"})
            return
        self._send(404, {"detail": "not found"})

    def do_PUT(self) -> None:
        length = int(self.headers.get("Content-Length", "0"))
        body = self.rfile.read(length)
        parts = self.path.strip("/").split("/")
        # internal/uploads/{id}/chunks/{index}
        upload_id = parts[2]
        index = int(parts[4])
        CHUNKS.setdefault(upload_id, {})[index] = body
        self._send(200, {"ok": True})


if __name__ == "__main__":
    server = ThreadingHTTPServer(("127.0.0.1", 19081), Handler)
    print("fake server2 listening on http://127.0.0.1:19081")
    server.serve_forever()
